File: src/positioning/libqgeocoordinate.cpp
#include <QGeoCoordinate>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qgeocoordinate.h>
#include "libqgeocoordinate.h"
#include "libqgeocoordinate.hxx"

QGeoCoordinate* QGeoCoordinate_new() {
    return new QGeoCoordinate();
}

QGeoCoordinate* QGeoCoordinate_new2(double latitude, double longitude) {
    return new QGeoCoordinate(static_cast<double>(latitude), static_cast<double>(longitude));
}

QGeoCoordinate* QGeoCoordinate_new3(double latitude, double longitude, double altitude) {
    return new QGeoCoordinate(static_cast<double>(latitude), static_cast<double>(longitude), static_cast<double>(altitude));
}

QGeoCoordinate* QGeoCoordinate_new4(const QGeoCoordinate* other) {
    return new QGeoCoordinate(*other);
}

void QGeoCoordinate_OperatorAssign(QGeoCoordinate* self, const QGeoCoordinate* other) {
    self->operator=(*other);
}

void QGeoCoordinate_Swap(QGeoCoordinate* self, QGeoCoordinate* other) {
    self->swap(*other);
}

bool QGeoCoordinate_IsValid(const QGeoCoordinate* self) {
    return self->isValid();
}

int QGeoCoordinate_Type(const QGeoCoordinate* self) {
    return static_cast<int>(self->type());
}

void QGeoCoordinate_SetLatitude(QGeoCoordinate* self, double latitude) {
    self->setLatitude(static_cast<double>(latitude));
}

double QGeoCoordinate_Latitude(const QGeoCoordinate* self) {
    return self->latitude();
}

void QGeoCoordinate_SetLongitude(QGeoCoordinate* self, double longitude) {
    self->setLongitude(static_cast<double>(longitude));
}

double QGeoCoordinate_Longitude(const QGeoCoordinate* self) {
    return self->longitude();
}

void QGeoCoordinate_SetAltitude(QGeoCoordinate* self, double altitude) {
    self->setAltitude(static_cast<double>(altitude));
}

double QGeoCoordinate_Altitude(const QGeoCoordinate* self) {
    return self->altitude();
}

double QGeoCoordinate_DistanceTo(const QGeoCoordinate* self, const QGeoCoordinate* other) {
    return static_cast<double>(self->distanceTo(*other));
}

double QGeoCoordinate_AzimuthTo(const QGeoCoordinate* self, const QGeoCoordinate* other) {
    return static_cast<double>(self->azimuthTo(*other));
}

QGeoCoordinate* QGeoCoordinate_AtDistanceAndAzimuth(const QGeoCoordinate* self, double distance, double azimuth) {
    return new QGeoCoordinate(self->atDistanceAndAzimuth(static_cast<qreal>(distance), static_cast<qreal>(azimuth)));
}

libqt_string QGeoCoordinate_ToString(const QGeoCoordinate* self) {
    QString _ret = self->toString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

QGeoCoordinate* QGeoCoordinate_AtDistanceAndAzimuth3(const QGeoCoordinate* self, double distance, double azimuth, double distanceUp) {
    return new QGeoCoordinate(self->atDistanceAndAzimuth(static_cast<qreal>(distance), static_cast<qreal>(azimuth), static_cast<qreal>(distanceUp)));
}

libqt_string QGeoCoordinate_ToString1(const QGeoCoordinate* self, int format) {
    QString _ret = self->toString(static_cast<QGeoCoordinate::CoordinateFormat>(format));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QGeoCoordinate_Delete(QGeoCoordinate* self) {
    delete self;
}

File: src/positioning/libqgeocoordinate.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qgeocoordinate_enums = enums;
const std = @import("std");

/// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html)
pub const QGeoCoordinate = extern struct {
    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.QGeoCoordinate,

    pub const _is_QGeoCoordinate = {};

    /// New constructs a new QGeoCoordinate object.
    ///
    pub fn New() QGeoCoordinate {
        return .{ .ptr = qtc.QGeoCoordinate_new() };
    }

    /// New2 constructs a new QGeoCoordinate object.
    ///
    /// ## Parameter(s):
    ///
    /// ` latitude: f64 `
    ///
    /// ` longitude: f64 `
    ///
    pub fn New2(latitude: f64, longitude: f64) QGeoCoordinate {
        return .{ .ptr = qtc.QGeoCoordinate_new2(@bitCast(latitude), @bitCast(longitude)) };
    }

    /// New3 constructs a new QGeoCoordinate object.
    ///
    /// ## Parameter(s):
    ///
    /// ` latitude: f64 `
    ///
    /// ` longitude: f64 `
    ///
    /// ` altitude: f64 `
    ///
    pub fn New3(latitude: f64, longitude: f64, altitude: f64) QGeoCoordinate {
        return .{ .ptr = qtc.QGeoCoordinate_new3(@bitCast(latitude), @bitCast(longitude), @bitCast(altitude)) };
    }

    /// New4 constructs a new QGeoCoordinate object.
    ///
    /// ## Parameter(s):
    ///
    /// ` other: QGeoCoordinate `
    ///
    pub fn New4(other: anytype) QGeoCoordinate {
        comptime _ = @TypeOf(other)._is_QGeoCoordinate;
        return .{ .ptr = qtc.QGeoCoordinate_new4(@ptrCast(other.ptr)) };
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#operator-eq)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` other: QGeoCoordinate `
    ///
    pub fn OperatorAssign(self: QGeoCoordinate, other: anytype) void {
        comptime _ = @TypeOf(other)._is_QGeoCoordinate;
        qtc.QGeoCoordinate_OperatorAssign(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#swap)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` other: QGeoCoordinate `
    ///
    pub fn Swap(self: QGeoCoordinate, other: anytype) void {
        comptime _ = @TypeOf(other)._is_QGeoCoordinate;
        qtc.QGeoCoordinate_Swap(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#isValid)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    pub fn IsValid(self: QGeoCoordinate) bool {
        return qtc.QGeoCoordinate_IsValid(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#type)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ## Returns:
    ///
    /// ` qgeocoordinate_enums.CoordinateType `
    ///
    pub fn Type(self: QGeoCoordinate) i32 {
        return qtc.QGeoCoordinate_Type(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#setLatitude)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` latitude: f64 `
    ///
    pub fn SetLatitude(self: QGeoCoordinate, latitude: f64) void {
        qtc.QGeoCoordinate_SetLatitude(@ptrCast(self.ptr), @bitCast(latitude));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#latitude)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    pub fn Latitude(self: QGeoCoordinate) f64 {
        return qtc.QGeoCoordinate_Latitude(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#setLongitude)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` longitude: f64 `
    ///
    pub fn SetLongitude(self: QGeoCoordinate, longitude: f64) void {
        qtc.QGeoCoordinate_SetLongitude(@ptrCast(self.ptr), @bitCast(longitude));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#longitude)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    pub fn Longitude(self: QGeoCoordinate) f64 {
        return qtc.QGeoCoordinate_Longitude(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#setAltitude)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` altitude: f64 `
    ///
    pub fn SetAltitude(self: QGeoCoordinate, altitude: f64) void {
        qtc.QGeoCoordinate_SetAltitude(@ptrCast(self.ptr), @bitCast(altitude));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#altitude)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    pub fn Altitude(self: QGeoCoordinate) f64 {
        return qtc.QGeoCoordinate_Altitude(@ptrCast(self.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#distanceTo)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` other: QGeoCoordinate `
    ///
    pub fn DistanceTo(self: QGeoCoordinate, other: anytype) f64 {
        comptime _ = @TypeOf(other)._is_QGeoCoordinate;
        return qtc.QGeoCoordinate_DistanceTo(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#azimuthTo)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` other: QGeoCoordinate `
    ///
    pub fn AzimuthTo(self: QGeoCoordinate, other: anytype) f64 {
        comptime _ = @TypeOf(other)._is_QGeoCoordinate;
        return qtc.QGeoCoordinate_AzimuthTo(@ptrCast(self.ptr), @ptrCast(other.ptr));
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#atDistanceAndAzimuth)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` distance: f64 `
    ///
    /// ` azimuth: f64 `
    ///
    pub fn AtDistanceAndAzimuth(self: QGeoCoordinate, distance: f64, azimuth: f64) QGeoCoordinate {
        return .{ .ptr = qtc.QGeoCoordinate_AtDistanceAndAzimuth(@ptrCast(self.ptr), @bitCast(distance), @bitCast(azimuth)) };
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#toString)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn ToString(self: QGeoCoordinate, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.QGeoCoordinate_ToString(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qgeocoordinate.ToString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#atDistanceAndAzimuth)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` distance: f64 `
    ///
    /// ` azimuth: f64 `
    ///
    /// ` distanceUp: f64 `
    ///
    pub fn AtDistanceAndAzimuth3(self: QGeoCoordinate, distance: f64, azimuth: f64, distanceUp: f64) QGeoCoordinate {
        return .{ .ptr = qtc.QGeoCoordinate_AtDistanceAndAzimuth3(@ptrCast(self.ptr), @bitCast(distance), @bitCast(azimuth), @bitCast(distanceUp)) };
    }

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#toString)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: QGeoCoordinate `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` format: qgeocoordinate_enums.CoordinateFormat `
    ///
    pub fn ToString1(self: QGeoCoordinate, allocator: std.mem.Allocator, format: i32) []const u8 {
        var _str = qtc.QGeoCoordinate_ToString1(@ptrCast(self.ptr), @bitCast(format));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qgeocoordinate.ToString1: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `Delete` instead
    ///
    pub const QDelete = Delete;

    /// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#dtor.QGeoCoordinate)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ## Parameter:
    ///
    /// ` self: QGeoCoordinate `
    ///
    pub fn Delete(self: QGeoCoordinate) void {
        qtc.QGeoCoordinate_Delete(@ptrCast(self.ptr));
    }
};

/// ### [Upstream resources](https://doc.qt.io/qt-6/qgeocoordinate.html#public-types)
pub const enums = struct {
    pub const CoordinateType = enum(i32) {
        pub const InvalidCoordinate: i32 = 0;
        pub const Coordinate2D: i32 = 1;
        pub const Coordinate3D: i32 = 2;
    };

    pub const CoordinateFormat = enum(i32) {
        pub const Degrees: i32 = 0;
        pub const DegreesWithHemisphere: i32 = 1;
        pub const DegreesMinutes: i32 = 2;
        pub const DegreesMinutesWithHemisphere: i32 = 3;
        pub const DegreesMinutesSeconds: i32 = 4;
        pub const DegreesMinutesSecondsWithHemisphere: i32 = 5;
    };
};
